File: 2-serverless-java/src/main/java/com/github/moravcik/configtracker/cdk/NotificationsNestedStack.java
package com.github.moravcik.configtracker.cdk;

import org.jetbrains.annotations.NotNull;
import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.eventsources.SqsEventSource;
import software.amazon.awscdk.services.sns.ITopic;
import software.amazon.awscdk.services.sns.subscriptions.SqsSubscription;
import software.amazon.awscdk.services.sqs.Queue;
import software.constructs.Construct;

import java.util.Map;

import static com.github.moravcik.configtracker.cdk.ConfigTrackerApp.resourcePrefix;
import static com.github.moravcik.configtracker.cdk.LambdaUtils.*;

public class NotificationsNestedStack extends NestedStack {

    public NotificationsNestedStack(@NotNull Construct scope, @NotNull String id, @NotNull ITopic configChangesTopic) {
        super(scope, id);

        Queue notificationsQueue = Queue.Builder.create(this, "NotificationsQueue")
                .queueName(resourcePrefix + "-config-changes-queue")
                .build();

        configChangesTopic.addSubscription(SqsSubscription.Builder.create(notificationsQueue)
                .rawMessageDelivery(true)
                .build());

        Function configNotificationHandler = createLambdaFunctionBuilder(this, "ConfigNotificationHandler")
                .environment(mergeEnvironment(
                        BASE_LAMBDA_ENVIRONMENT,
                        Map.of("SPRING_CLOUD_FUNCTION_DEFINITION", "configNotificationSqsHandler")))
                .build();

        configNotificationHandler.addEventSource(SqsEventSource.Builder.create(notificationsQueue).build());
    }
}

File: 1-serverless-ts/cdk/notifications.nested-stack.ts
import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { ITopic } from 'aws-cdk-lib/aws-sns';
import { SqsSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { Construct } from 'constructs';
import { baseLambdaProps, resourcePrefix } from './config-tracker.app';

export class NotificationsNestedStack extends NestedStack {
  constructor(scope: Construct, id: string, props: NestedStackProps & { configChangesTopic: ITopic }) {
    super(scope, id, props);

    const notificationsQueue = new Queue(this, 'NotificationsQueue', {queueName: `${resourcePrefix}-config-changes-queue`});
    props.configChangesTopic.addSubscription(new SqsSubscription(notificationsQueue, {rawMessageDelivery: true}));

    const configNotificationHandler = new NodejsFunction(this, 'ConfigNotificationHandler', {
      entry: 'lib/lambda/config-notification-sqs.handler.ts',
      ...baseLambdaProps
    });
    configNotificationHandler.addEventSource(new SqsEventSource(notificationsQueue));
  }
}